### PR TITLE
test: add coverage for feebumper uncomputable cluster error path

### LINKED ISSUE #34902: Add test coverage for bump fee calculation failure in feebumper
### Motivation

There is no test coverage for the error here:
https://github.com/bitcoin/bitcoin/blob/1a1f584360ce053f9f832e310d063d686f86c6d7/src/wallet/feebumper.cpp#L84-L86

Lack of coverage can be seen for line 85 (currently) here: https://maflcko.github.io/b-c-cov/total.coverage/src/wallet/feebumper.cpp.gcov.html

### Possible solution

A functional test using the bumpfee RPC can be a possible solution.
There is a wallet_bumpfee test file that has a good amunt of tests, which can be referred to come up with a solution.

### Useful Skills

* Compiling Bitcoin Core from source
* Running the C++ unit tests and the Python functional tests
* Wallet transaction creation
* Handling unconfirmed UTXOs while creating and bumping wallet transactions
* Fee rate and mempool fee environment theoretical knowledge

### Guidance for new contributors

Want to work on this issue?

For guidance on contributing, please read [CONTRIBUTING.md](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md) before opening your pull request.

### PR DESCRIPTION
**Pull request by 151henry151**
Add functional test coverage for the error path in `CheckFeeRate()` when `calculateCombinedBumpFee()` returns `nullopt` because unconfirmed inputs span more than 500 mempool transactions (`CTxMemPool::GatherClusters` DoS limit).

The crash on this path was fixed in #34870; this PR adds the missing test requested in #34902.

The test builds eight independent 64-transaction clusters (512 total, above the 500 limit), keeps a wallet-owned RBF transaction that spends one output from each cluster out of the mempool, and asserts that `bumpfee` with an explicit `fee_rate` returns the expected wallet error instead of proceeding with an empty optional.

Closes #34902.

### CODE REVIEW

**Comment by 151henry151** [comment]
@furszy Would you mind taking a look when you have a chance? This adds the functional test coverage for the error path you fixed in #34870.

`CheckFeeRate()` only runs when the caller passes an explicit `fee_rate`; default `bumpfee` uses fee estimation and does not hit this branch, which is why the test passes `{"fee_rate": NORMAL}`.

File: test/functional/wallet_bumpfee.py [hunk 1]
```diff
@@ -829,5 +838,122 @@ def test_bumpfee_with_feerate_ignores_walletincrementalrelayfee(self, rbf_node,
...
+                pass
+        self.generate(rbf_node, 1, sync_fun=self.no_op)
+        self.restart_node(1, self.extra_args[1])
+        rbf_node.walletpassphrase(WALLET_PASSPHRASE, WALLET_PASSPHRASE_TIMEOUT)
+        self.connect_nodes(1, 0)
+        self.sync_all()
+        excess = rbf_node.getbalance() - balance_before
+        if excess > 0:
+            rbf_node.sendtoaddress(peer_node.getnewaddress(), excess)
+        self.clear_mempool()
```
Comment (furszy) [inline test/functional/wallet_bumpfee.py]: I think you could simplify this:

```python3
def test_bumpfee_uncomputable_cluster(self, rbf_node, dest_address):
    self.log.info('Test that bumpfee fails when unconfirmed UTXOs depend on an enormous cluster')
    # CheckFeeRate() can only fail to compute a combined bump fee when
    # CTxMemPool::GatherClusters() hits its 500 transaction limit. Clusters are capped
    # at 64 transactions, so we need the tx we bump to spend from 8 of them.
    #
    # That tx can't be in the mempool once they're full, or it would merge all 8 into
    # one and get rejected. We can't create it then either, since coin selection runs
    # into the same MiniMiner limit. So we broadcast it while the clusters are still
    # single transactions and evict it afterwards by raising -minrelaytxfee. It stays
    # in the wallet, which is all bumpfee needs.
    NUM_CLUSTERS = 8             # 8 * 64 = 512, above the 500 gather limit
    CLUSTER_SIZE = 64
    SEED_PAYMENT = COIN // 100   # paid to rbf_node by every seed
    SEED_FEE = 20_000            # high enough to survive the -minrelaytxfee bump below
    ORIGINAL_TX_FEE = 2000       # ~3 sat/vB, low enough to get evicted by it

    # Keep the peer disconnected so we don't relay txs to it.
    self.disconnect_nodes(0, 1)

    # Fund a MiniWallet on rbf_node so the cluster transactions end up in its mempool,
    # which is the one the bump fee calculation looks at. The tag keeps these coins
    # separate from the default MiniWallet other tests use on this node. get_utxo()
    # skips immature coinbases, so mine COINBASE_MATURITY blocks on top.
    miniwallet = MiniWallet(rbf_node, tag_name="feebumper_cluster")
    self.generatetoaddress(rbf_node, COINBASE_MATURITY + NUM_CLUSTERS,
                           miniwallet.get_address(), sync_fun=self.no_op)
    miniwallet.rescan_utxos()

    # Each seed pays rbf_node and keeps a MiniWallet output we grow the cluster from.
    rbf_spk = bytes.fromhex(rbf_node.getaddressinfo(rbf_node.getnewaddress())["scriptPubKey"])
    seeds = [miniwallet.send_to(from_node=rbf_node, scriptPubKey=rbf_spk,
                                amount=SEED_PAYMENT, fee=SEED_FEE)
             for _ in range(NUM_CLUSTERS)]

    # Broadcast the tx we'll bump while every cluster is still a single transaction.
    inputs = [{"txid": seed["txid"], "vout": seed["sent_vout"]} for seed in seeds]
    outputs = {dest_address: Decimal(NUM_CLUSTERS * SEED_PAYMENT - ORIGINAL_TX_FEE) / COIN}
    signed = rbf_node.signrawtransactionwithwallet(rbf_node.createrawtransaction(inputs, outputs))
    original_txid = rbf_node.sendrawtransaction(signed["hex"])

    # Bumping -minrelaytxfee to 10 sat/vB drops it on mempool reload and rejects the
    # resubmission the wallet does on startup. The seeds pay much more, so they stay.
    # Note the bump below uses NORMAL (100 sat/vB); if that ever ends up under the
    # -minrelaytxfee here, CheckFeeRate() bails on its mempool minimum check instead.
    self.restart_node(1, ["-minrelaytxfee=0.0001"] + self.extra_args[1])
    rbf_node.walletpassphrase(WALLET_PASSPHRASE, WALLET_PASSPHRASE_TIMEOUT)
    assert_equal(set(rbf_node.getrawmempool()), {seed["txid"] for seed in seeds})
    assert_equal(rbf_node.gettransaction(original_txid)["confirmations"], 0)

    # Grow each seed into a full cluster. MiniWallet's default 300 sat/vB keeps these
    # above the raised -minrelaytxfee.
    for seed in seeds:
        miniwallet.send_self_transfer_chain(from_node=rbf_node,
                                            chain_length=CLUSTER_SIZE - 1,
                                            utxo_to_spend=miniwallet.get_utxo(txid=seed["txid"]))
    # Make sure the setup worked, so the check below can't pass for the wrong reason.
    assert_equal(len(rbf_node.getrawmempool()), NUM_CLUSTERS * CLUSTER_SIZE)

    assert_raises_rpc_error(-4,
                            "Failed to calculate bump fees, because unconfirmed UTXOs depend on an enormous cluster of unconfirmed transactions.",
                            rbf_node.bumpfee, original_txid, {"fee_rate": NORMAL})

    # Restore initial state
    rbf_node.abandontransaction(original_txid)
    self.generate(rbf_node, 1, sync_fun=self.no_op)
    self.restart_node(1, self.extra_args[1])
    rbf_node.walletpassphrase(WALLET_PASSPHRASE, WALLET_PASSPHRASE_TIMEOUT)
    self.connect_nodes(1, 0)
    self.sync_all()
```

**Comment by 151henry151** [comment]
Pushed in `47632ab` — applied your simplification (squashed into the original commit). Thanks for the review.

**Review by furszy — COMMENTED** [review]
ACK 47632ab2a3876825576b3f07e6f80e799e6ed58f

Tested it fails without the merged fix commit 6072a2a6a1fdde8444e3515f0d201de3fb3923a1

**Comment by rapennas** [comment]
Tested ACK 47632ab2a3

Tested on Linux with GCC 13.3.0, based on master 8a4bab8e97.

The functional test passed as proposed:

```bash
./build/test/functional/wallet_bumpfee.py
```

I also tested the `GatherClusters()` boundary by locally changing the cluster sizes:

- With 501 transactions (`[64] * 7 + [53]`), `bumpfee` returned the expected `Failed to calculate bump fees...` error.
- With 500 transactions (`[64] * 7 + [52]`), the cluster gathering step succeeded and transaction creation subsequently returned `Unable to create transaction. too many unconfirmed transactions in cluster`.

After updating the expected error for the local 500-transaction experiment, the functional test passed. This confirms that 500 transactions are still computable, while 501 is the first value that triggers the error path covered by this PR.

**Review by sedited — APPROVED** [review]
ACK 47632ab2a3876825576b3f07e6f80e799e6ed58f
